File: app/src/main/java/dev/ragnarok/fenrir/fragment/ProxyManagerFrgament.java
package dev.ragnarok.fenrir.fragment;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.ProxiesAdapter;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.model.ProxyConfig;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.ProxyManagerPresenter;
import dev.ragnarok.fenrir.mvp.view.IProxyManagerView;
import dev.ragnarok.fenrir.place.PlaceFactory;

public class ProxyManagerFrgament extends BaseMvpFragment<ProxyManagerPresenter, IProxyManagerView>
        implements IProxyManagerView, ProxiesAdapter.ActionListener {

    private ProxiesAdapter mProxiesAdapter;

    public static ProxyManagerFrgament newInstance() {
        Bundle args = new Bundle();
        ProxyManagerFrgament fragment = new ProxyManagerFrgament();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_proxy_manager, container, false);

        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));

        RecyclerView recyclerView = root.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireActivity()));

        mProxiesAdapter = new ProxiesAdapter(Collections.emptyList(), this);
        recyclerView.setAdapter(mProxiesAdapter);
        return root;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.proxies, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_add) {
            callPresenter(ProxyManagerPresenter::fireAddClick);
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @NonNull
    @Override
    public IPresenterFactory<ProxyManagerPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> new ProxyManagerPresenter(saveInstanceState);
    }

    @Override
    public void displayData(List<ProxyConfig> configs, ProxyConfig active) {
        if (nonNull(mProxiesAdapter)) {
            mProxiesAdapter.setData(configs, active);
        }
    }

    @Override
    public void notifyItemAdded(int position) {
        if (nonNull(mProxiesAdapter)) {
            mProxiesAdapter.notifyItemInserted(position + mProxiesAdapter.getHeadersCount());
        }
    }

    @Override
    public void notifyItemRemoved(int position) {
        if (nonNull(mProxiesAdapter)) {
            mProxiesAdapter.notifyItemRemoved(position + mProxiesAdapter.getHeadersCount());
        }
    }

    @Override
    public void setActiveAndNotifyDataSetChanged(ProxyConfig config) {
        if (nonNull(mProxiesAdapter)) {
            mProxiesAdapter.setActive(config);
        }
    }

    @Override
    public void goToAddingScreen() {
        PlaceFactory.getProxyAddPlace().tryOpenWith(requireActivity());
    }

    @Override
    public void onDeleteClick(ProxyConfig config) {
        callPresenter(p -> p.fireDeleteClick(config));
    }

    @Override
    public void onSetAtiveClick(ProxyConfig config) {
        callPresenter(p -> p.fireActivateClick(config));
    }

    @Override
    public void onDisableClick(ProxyConfig config) {
        callPresenter(p -> p.fireDisableClick(config));
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/fragment/ProxyManagerFrgament.kt
package dev.ragnarok.fenrir.fragment

import android.os.Bundle
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.adapter.ProxiesAdapter
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.model.ProxyConfig
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory
import dev.ragnarok.fenrir.mvp.presenter.ProxyManagerPresenter
import dev.ragnarok.fenrir.mvp.view.IProxyManagerView
import dev.ragnarok.fenrir.place.PlaceFactory.proxyAddPlace

class ProxyManagerFrgament : BaseMvpFragment<ProxyManagerPresenter, IProxyManagerView>(),
    IProxyManagerView, ProxiesAdapter.ActionListener {
    private var mProxiesAdapter: ProxiesAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = inflater.inflate(R.layout.fragment_proxy_manager, container, false)
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        val recyclerView: RecyclerView = root.findViewById(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(requireActivity())
        mProxiesAdapter = ProxiesAdapter(mutableListOf(), this)
        recyclerView.adapter = mProxiesAdapter
        return root
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.proxies, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_add) {
            presenter?.fireAddClick()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<ProxyManagerPresenter> {
        return object : IPresenterFactory<ProxyManagerPresenter> {
            override fun create(): ProxyManagerPresenter {
                return ProxyManagerPresenter(saveInstanceState)
            }
        }
    }

    override fun displayData(configs: MutableList<ProxyConfig>, active: ProxyConfig?) {
        mProxiesAdapter?.setData(configs, active)
    }

    override fun notifyItemAdded(position: Int) {
        mProxiesAdapter?.notifyItemInserted(position + (mProxiesAdapter?.headersCount ?: 0))
    }

    override fun notifyItemRemoved(position: Int) {
        mProxiesAdapter?.notifyItemRemoved(position + (mProxiesAdapter?.headersCount ?: 0))
    }

    override fun setActiveAndNotifyDataSetChanged(config: ProxyConfig?) {
        mProxiesAdapter?.setActive(config)
    }

    override fun goToAddingScreen() {
        proxyAddPlace.tryOpenWith(requireActivity())
    }

    override fun onDeleteClick(config: ProxyConfig) {
        presenter?.fireDeleteClick(
            config
        )
    }

    override fun onSetAtiveClick(config: ProxyConfig) {
        presenter?.fireActivateClick(
            config
        )
    }

    override fun onDisableClick(config: ProxyConfig) {
        presenter?.fireDisableClick()
    }

    companion object {
        fun newInstance(): ProxyManagerFrgament {
            val args = Bundle()
            val fragment = ProxyManagerFrgament()
            fragment.arguments = args
            return fragment
        }
    }
}